### PR TITLE
Replaced resample function by new implementation (2)

### PR DESCRIPTION
**Pull request by AlexFuster**
#1392

### CODE REVIEW

**Comment by yongtang** [comment]
@AlexFuster if you have black installed,  can you just run black audio_ops.py and see if the file is formatted automatically?

**Comment by AlexFuster** [comment]
The tests seem to be taking too long

**Comment by AlexFuster** [comment]
ok, here we go again

**Comment by AlexFuster** [comment]
Oh yeah, I forgot to mention It does just support floating point audios. Is this a problem?

**Comment by yongtang** [comment]
> Oh yeah, I forgot to mention It does just support floating point audios. Is this a problem?

@AlexFuster The existing implementation support both int16 and float32. Though I think it will not be too difficult to add int16 support by casting to float32 and back.

**Comment by yongtang** [comment]
@AlexFuster The test failures are legitimate. The reason being
1. `Tensor` in tensorflow may not be treated as python native value (e.g, `1`, `1.1`, etc). Things like `int(x)` may not work when `x` is a tensor. Instead, an explicit op of `tf.cast(x, tf.int32)` is needed.
2. A `tensor`'s shape may not be known statically before hand. So `x.shape` could just return `None` in case shape is not known yet.
3. When `x` is a tensor, the `if...else...` logic does not work as tensor's value is not statically available (before run). Instead a TF specific op of `tf.case` or `tf.cond` has to be used (you might notice the original implementation consists of `tf.case` to replace `if...elif...else`.

You may want to take a look at the test failures and see if some change is needed. I will also take a look.

**Comment by AlexFuster** [comment]
I don't think the error comes from any of those 3 problems. It is just `tf.nn.conv1d` not accepting `int16` values as inputs. I'll try adding int16 support by casting and casting back as you said

**Comment by yongtang** [comment]
With `int16` tests may pass, though in the original implementation the resample actually support passing `rate_in`, `rate_out` as tensor and input with unknown shape. I think with new implementation this ability is removed as I could see `gcd` and `int(rate_in)` will cause the failure if a tensor with unknown rank is passed for rate_in. The input may also cause some issue if it is a tensor with unknown shape. We can get the tests pass and address the rest in follow up PRs though.

**Comment by AlexFuster** [comment]
Why would you want a tensor with a shape other than () being passed as `rate_in`?

**Comment by yongtang** [comment]
> Why would you want a tensor with a shape other than () being passed as rate_in?
@AlexFuster Sorry it should be  `a tensor without constant value`. In graph mode (such as within tf.data.Dataset pipeline), user may pass `input`, `rate_in`, `rate_out` as tensor without const value. In that case, the graph mode will fail. They can be avoided by using ops that will work for tensor. For example, `int(rate_in)` can be converted to use `tf.cast(rate_in, tf.int32)`, `math.gcd` can be replaced with ops as `tf.experimental.numpy.gcd` I think.

**Comment by AlexFuster** [comment]
There's a problem `tf.nn.conv1d` does not accept a tensor as a value for stride. So the operation `resampled = tf.nn.conv1d(waveform, kernel, stride=rate_in, padding="VALID")` is nor working for me when `rate_in` is a tensor. It does if I use `rate_in.numpy()`, but that would not be compatible with graph mode

**Comment by AlexFuster** [comment]
@googlebot I fixed it.

**Review by rasul247 — APPROVED** [review]
